crates/rpc-types-engine/src/payload/mod.rs

```diff
@@ -0,0 +1,21 @@
+//! Versioned Optimism execution payloads
+
+pub mod v3;
+pub mod v4;
+
+use crate::OpExecutionPayloadV4;
+use alloy_rpc_types_engine::{ExecutionPayloadV2, ExecutionPayloadV3};
+
+/// An execution payload, which can be either [`ExecutionPayloadV2`], [`ExecutionPayloadV3`], or
+/// [`OpExecutionPayloadV4`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(untagged))]
+pub enum OpExecutionPayload {
+    /// V2 payload
+    V2(ExecutionPayloadV2),
+    /// V3 payload
+    V3(ExecutionPayloadV3),
+    /// V4 payload
+    V4(OpExecutionPayloadV4),
```

Comment:
lgtm,

one thing that could be problematic, perhaps even for L1 types is that untagged here doesn't quite work for V4 because this could already succeed on v3, 
for the engineapi this isn't problematic, but we should track manual deserde for this